evm/src/main/java/org/hyperledger/besu/evm/operation/SelfDestructOperation.java

```diff
@@ -60,19 +75,22 @@ public OperationResult execute(final MessageFrame frame, final EVM evm) {
     final Address address = frame.getRecipientAddress();
     final MutableAccount account = frame.getWorldUpdater().getAccount(address).getMutable();
 
-    frame.addSelfDestruct(address);
+    if (!eip6780Semantics || frame.wasCreatedInTransaction(address)) {
+      frame.addSelfDestruct(address);
+    }
 
     final MutableAccount recipient =
         frame.getWorldUpdater().getOrCreate(recipientAddress).getMutable();
 
-    if (!account.getAddress().equals(recipient.getAddress())) {
-      recipient.incrementBalance(account.getBalance());
+    Wei contractBalance = account.getBalance();
+    if (eip6780Semantics || !account.getAddress().equals(recipient.getAddress())) {
+      recipient.incrementBalance(contractBalance);
     }
 
     // add refund in message frame
-    frame.addRefund(recipient.getAddress(), account.getBalance());
+    frame.addRefund(recipient.getAddress(), contractBalance);
 
-    account.setBalance(Wei.ZERO);
+    account.decrementBalance(contractBalance);
```

Comment:
Might be worth a comment to explain the case where this balance should not be set to zero, as before.

Seems like for most cases this is equivalent to setting to zero, except the case where the contract account being "hollowed" (because it was created in a prior transaction) had an existing balance in it...IIUC, this is a rare and specific usage of CREATE2?

Reply:
Worse than that, there was a bug relating to repeated self-destructs of the same address.  I rewrote the logic into something much simpler.